src/psyclone/psyir/symbols/typed_symbol.py

```diff
@@ -0,0 +1,205 @@
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors R. W. Ford, A. R. Porter and S. Siso, STFC Daresbury Lab
+#         I. Kavcic, Met Office
+#         J. Henrichs, Bureau of Meteorology
+# -----------------------------------------------------------------------------
+
+'''This module contains the TypedSymbol class.
+
+'''
+
+from __future__ import absolute_import
+import abc
+import six
+from psyclone.psyir.symbols.data_type_symbol import DataTypeSymbol
+from psyclone.psyir.symbols.symbol import Symbol
+
+
+@six.add_metaclass(abc.ABCMeta)
+class TypedSymbol(Symbol):
+    '''
+    Abstract base class for those Symbols that have an associated datatype.
+
+    :param str name: name of the symbol.
+    :param datatype: data type of the symbol.
+    :type datatype: :py:class:`psyclone.psyir.symbols.DataType`
+    :param visibility: the visibility of the symbol.
+    :type visibility: :py:class:`psyclone.psyir.symbols.Symbol.Visibility`
+    :param interface: optional object describing the interface to this \
+        symbol (i.e. whether it is passed as a routine argument or \
+        accessed in some other way). Defaults to \
+        :py:class:`psyclone.psyir.symbols.LocalInterface`
+    :type interface: :py:class:`psyclone.psyir.symbols.symbol.SymbolInterface`
+
+    :raises TypeError: if the name is not a str or visibility is not an \
+                       instance of Visibility.
+
+    '''
+    def __init__(self, name, datatype, visibility=Symbol.DEFAULT_VISIBILITY,
+                 interface=None):
+
+        super(TypedSymbol, self).__init__(name, visibility=visibility,
+                                          interface=interface)
+        # The following attribute has a setter method (with error checking)
+        self._datatype = None
+        self.datatype = datatype
+
+    @abc.abstractmethod
+    def __str__(self):
+        ''' Return a str describing this object. '''
+
+    @property
+    def datatype(self):
+        '''
+        :returns: datatype of the TypedSymbol.
+        :rtype: str
+        '''
+        return self._datatype
+
+    @datatype.setter
+    def datatype(self, value):
+        ''' Setter for the datatype of a TypedSymbol.
+
+        :param value: new value for datatype.
+        :type value: :py:class:`psyclone.psyir.symbols.DataType` or \
+                     :py:class:`psyclone.psyir.symbols.DataTypeSymbol`
+
+        :raises TypeError: if value is not of the correct type.
+        :raises NotImplementedError: if the specified data type is invalid.
```

Comment:
Remove docstring of NotImplementedError and check if the top-level import is still not possible.

Reply:
docstring fixed and import is still required :-(